src/ivectorbin/ivector-extract-dense.cc

```diff
@@ -80,107 +140,142 @@ int main(int argc, char *argv[]) {
-        continue;
-      }
-      const Matrix<BaseFloat> &mat = feature_reader.Value();
-      Posterior posterior = posterior_reader.Value(utt);
+    TokenVectorWriter ivector_ranges_writer(ivector_ranges_wspecifier);
+    TokenVectorWriter ivector_weights_writer(ivector_weights_wspecifier);
+
+    for (unordered_map<std::string, std::vector<Segments*> >::iterator iter = record2seg.begin();
+         iter != record2seg.end(); ++iter) {
+      std::vector<Matrix<BaseFloat> > ivector_list;
+      std::vector<std::string> ivector_ranges;
+      std::vector<std::string> ivector_weights;
+      int32 tot_size = 0;
+      for (int32 seg_num = 0; seg_num < iter->second.size(); seg_num++) {
+        std::string utt = iter->second[seg_num]->segment;
+        if (!posterior_reader.HasKey(utt)) {
+          KALDI_WARN << "No posteriors for utterance " << utt;
+          num_err++;
+          continue;
+        }
+        const Matrix<BaseFloat> &mat = feature_reader.Value(utt);
+        Posterior posterior = posterior_reader.Value(utt);
       
-      if (static_cast<int32>(posterior.size()) != mat.NumRows()) {
-        KALDI_WARN << "Size mismatch between posterior " << posterior.size()
-                   << " and features " << mat.NumRows() << " for utterance "
-                   << utt;
-        num_err++;
-        continue;
-      }
+        if (static_cast<int32>(posterior.size()) != mat.NumRows()) {
+          KALDI_WARN << "Size mismatch between posterior " << posterior.size()
+                     << " and features " << mat.NumRows() << " for utterance "
+                     << utt;
+          num_err++;
+          continue;
+        }
+
+        double *auxf_ptr = (compute_objf_change ? &tot_auxf_change : NULL );
 
-      double *auxf_ptr = (compute_objf_change ? &tot_auxf_change : NULL );
+        double this_t = opts.acoustic_weight * TotalPosterior(posterior),
+            max_count_scale = 1.0;
+        if (opts.max_count > 0 && this_t > opts.max_count) {
+          max_count_scale = opts.max_count / this_t;
+          KALDI_LOG << "Scaling stats for utterance " << utt << " by scale "
+                    << max_count_scale << " due to --max-count="
+                    << opts.max_count;
+          this_t = opts.max_count;
+        }
+        ScalePosterior(opts.acoustic_weight * max_count_scale,
+                       &posterior);
+        // note: now, this_t == sum of posteriors.
 
-      double this_t = opts.acoustic_weight * TotalPosterior(posterior),
-          max_count_scale = 1.0;
-      if (opts.max_count > 0 && this_t > opts.max_count) {
-        max_count_scale = opts.max_count / this_t;
-        KALDI_LOG << "Scaling stats for utterance " << utt << " by scale "
-                  << max_count_scale << " due to --max-count="
-                  << opts.max_count;
-        this_t = opts.max_count;
+        int32 num_chunks = std::max((int32)ceil((mat.NumRows() - chunk_size + period) / static_cast<BaseFloat>(period)), 1);
```

Comment:
Also, this line is doing casting in both the C style and C++ style. I think it will be more consistent to do the int32 casting using `static_cast<int32>()` instead of `(int32)`